test/functional/selfmanaged_apicast.go

```diff
@@ -0,0 +1,844 @@
+			Labels: map[string]string{
+				"app":                  "apicast",
+				"threescale_component": "apicast",
+			},
+			Annotations: map[string]string{
+				"openshift.io/host.generated": "true",
+			},
+		},
+		Spec: routev1.RouteSpec{
+			Port: &routev1.RoutePort{
+				TargetPort: intstr.FromString("proxy"),
+			},
+			TLS: &routev1.TLSConfig{
+				Termination: "edge",
+			},
+			To: routev1.RouteTargetReference{
+				Kind: "Service",
+				Name: "apicast-example-apicast",
+			},
+			Host: "selfmanaged-staging" + hostsub,
+		},
+	}
+	err := client.Create(ctx, route)
+	if err != nil {
+		log.Error("Error create Apicast Route", err)
+		return "", err
+	}
+	err = wait.Poll(time.Second*5, time.Minute*3, func() (done bool, err error) {
+		err = client.Get(ctx, k8sclient.ObjectKey{Name: route.Name, Namespace: route.Namespace}, route)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	})
+	routeHost := route.Spec.Host
+	log.Info("routeHost: " + routeHost)
+	return routeHost, nil
+}
+
+func customerLogin() error {
+	log.Info("Customer Login")
+	customerAdminUsername := fmt.Sprintf("%v%02d", defaultDedicatedAdminName, 1)
+	command := "oc login -u " + customerAdminUsername + " -p " + testcommon.DefaultPassword
+	err := runShellCommand(command)
+	if err != nil {
+		log.Error("Error login customer: ", err)
+		return err
+	}
+	return nil
+}
+
+func runShellCommand(command string) error {
+	cmd := exec.Command("bash", "-c", command)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+	if err != nil {
+		log.Error("Error: ", err)
+		return err
+	}
+	return nil
+}
+
+func promoteSelfManagedAPIcast(ctx context.Context, client k8sclient.Client, threeScaleAdminPortal string, token3scale string) (string, error) {
+	err := deleteManagedApicastRoutes(ctx, client)
+	if err != nil {
+		log.Error("Error deleting Managed Apicast Routes : ", err)
+		return "", err
+	}
+	serviceId, err := getServiceId(threeScaleAdminPortal, token3scale)
+	if err != nil {
+		log.Error("Error get Service ID", err)
+		return "", err
+	}
+	err = apiCastConfigPromote(threeScaleAdminPortal, token3scale, serviceId)
+	if err != nil {
+		log.Error("Error in 3scale api Proxy Config Promote : ", err)
+		return "", err
+	}
+	userKey, err := getUserKey(threeScaleAdminPortal, token3scale, serviceId)
+	if err != nil {
+		log.Error("Error get user key: ", err)
+		return "", err
+	}
+	return userKey, err
+}
+
+func validateDeploymentRequest(userKey, routeHost string) (int, error) {
+	httpRequest := "https://" + routeHost + "/?user_key=" + userKey
+	resp, err := http.Get(httpRequest)
+	if err != nil {
+		log.Error("HTTP Get error", err)
+		return 0, err
+	}
+	//log.Info("Response Code: " + strconv.Itoa(resp.StatusCode))
+	if resp.StatusCode == http.StatusOK {
+		defer resp.Body.Close()
+		_, err := ioutil.ReadAll(resp.Body) //bytes,err:= ...
+		if err != nil {
+			log.Error("Unable to read response body: ", err)
+			return 0, nil
+		}
+		return http.StatusOK, nil
```

Comment:
Fixed. It required cleanup. Before - i had print body here, but because of security issues, - print of body was removed (it had user key seems), and cleanup was not completed ... return nil - don't know ... maybe it was issue with logics, as response is 200 : ). 
Fixed, thank you